packages/@react-stately/list/src/useListState.ts

```diff
@@ -61,11 +61,35 @@ export function useListState<T extends object>(props: ListProps<T>): ListState<T
     , [collection, selectionState, layoutDelegate]
   );
 
+  useFocusedKeyReset(collection, selectionManager);
+
+  return {
+    collection,
+    disabledKeys,
+    selectionManager
+  };
+}
+
+/**
+ * Filters a collection using the provided filter function and returns a new ListState.
+ */
+export function useFilteredListState<T extends object>(state: ListState<T>, filterFn: ((nodeValue: string) => boolean) | null | undefined): ListState<T> {
+  let collection = useMemo(() => filterFn ? state.collection.filter!(filterFn) : state.collection, [state.collection, filterFn]);
+  let selectionManager = state.selectionManager.withCollection(collection);
```

Comment:
this seems easy to break, did we consider having the filtering happen inside layouts?
i guess then the keyboard delegates would need updating as well... hmmm, what other options do we have?

Reply:
what do you think would break?

Reply:
Mostly that it can be called anywhere and is a separate thing to remember to do
I'm not so much worried about it for ourselves, we only have to do it in a handful of places and they've all been taken care of in this pr, but for outside developers

Reply:
which thing are you talking about specifically? the selection manager thing or the entire hook?

Reply:
the selection manager specifically

Reply:
oh, yeah, the `useFilteredListState` is what you're supposed to use. I had to add the `withCollection` method to create a new selection manager for it because the original selection state and options needed to construct a SelectionManager aren't exposed (they are `private` in SelectionManager). So the alternative would be to make those public and then do `new SelectionManager` here. 🤷

Reply:
blargh, will resolve for now, will keep thinking on it